### ISSUE TITLE
TrueCrypt Keyfile Support for Yubikey NEO "function not supported"

### ISSUE THREAD

**Issue by tristan-k** [OpenSC/OpenSC#479]
[Disk encryption](https://github.com/OpenSC/OpenSC/wiki/Using-smart-cards-with-applications#disk-encryption)
> TrueCrypt can use PKCS#11 tokens as keyfile stores. NB! TrueCrypt does not use asymmetric keys generated on the card but stores symmetric keys as data files in the token! This requires write access to the token and keyfiles are extracted in plaintext on every use.

Quote from https://github.com/Yubico/ykneo-openpgp/issues/17#issuecomment-94499111

> Hello,
Now that I got a Yubikey Neo last generation, I would like to use the PIV applet with OpenSC for PKCS11 authentication into VeraCrypt (fork of TrueCrypt) to force the need of a Yubikey to decrypt my Volume.
My Yubikey & PIV applet is recognized by the OpenSC library that I specify in VeraCrypt.
However, when I try to use the opensc-pkcs11 library with Yubico inside VeraCrypt, I have an error saying "function not supported". I understand the problem come from OpenSC as it doesn't support objects like keyfiles.

![security token keyfiles_001](https://cloud.githubusercontent.com/assets/732025/8094163/873385c2-0fc5-11e5-89e2-cd8578d044b5.png)
![new security token keyfile properties_002](https://cloud.githubusercontent.com/assets/732025/8094162/87322948-0fc5-11e5-8ae4-cadb6e5653c3.png)
![truecrypt_003](https://cloud.githubusercontent.com/assets/732025/8094164/87345b82-0fc5-11e5-86d5-0104aa393303.png)

```
$ opensc-tool -a
Using reader with a card: Yubico Yubikey NEO OTP+CCID 00 00
```

Further reading:
http://forum.yubico.com/viewtopic.php?p=7166&sid=958cdb2de8b74e2320955704d3765a12#p7166
http://www.gossamer-threads.com/lists/gnupg/users/70510
http://sourceforge.net/p/veracrypt/discussion/features/thread/34fb4681/

**Comment by dengert** [comment]
On 6/10/2015 4:10 PM,  ǝʞɔoʃʞ uɐʇsıɹʇ wrote:
> Disk encryption <https://github.com/OpenSC/OpenSC/wiki/Using-smart-cards-with-applications#disk-encryption>
>
>     TrueCrypt can use PKCS#11 tokens as keyfile stores. NB! TrueCrypt does not use asymmetric keys generated on the card but stores symmetric keys as data files in the token! This requires write
>     access to the token and keyfiles are extracted in plaintext on every use.

The OpenSC support for the NEO is for the PIV applet on the NEO. PIV is defined by NIST 800-73-3 part 2:

   http://www.idmanagement.gov/resource/nist-special-publication-800-73-3-interfaces-piv-4-parts-piv-card-application-card-command

There are no symmetric key objects defined for PIV. There is no intention to add support to OpenSC for such an object, as it is not defined by NIST 800-73-3.

You may need to look for a different card or applet for the NEO that does support the storage and of a symmetric key
or use the PIV RSA encryption key to decrypt a symmetric key stored somewhere else.

BUT:

http://truecrypt.sourceforge.net/

Says:
WARNING: Using TrueCrypt is not secure as it may contain unfixed security issues
The development of TrueCrypt was ended in 5/2014 after Microsoft terminated support of Windows XP.


>
> Quote from Yubico/ykneo-openpgp#17 (comment) <https://github.com/Yubico/ykneo-openpgp/issues/17#issuecomment-94499111>
>
>     Hello,
>     Now that I got a Yubikey Neo last generation, I would like to use the PIV applet with OpenSC for PKCS11 authentication into VeraCrypt (fork of TrueCrypt) to force the need of a Yubikey to decrypt
>     my Volume.
>     My Yubikey & PIV applet is recognized by the OpenSC library that I specify in VeraCrypt.
>     However, when I try to use the opensc-pkcs11 library with Yubico inside VeraCrypt, I have an error saying "function not supported". I understand the problem come from OpenSC as it doesn't support
>     objects like keyfiles.
>
> security token keyfiles_001 <https://cloud.githubusercontent.com/assets/732025/8094163/873385c2-0fc5-11e5-89e2-cd8578d044b5.png>
> new security token keyfile properties_002 <https://cloud.githubusercontent.com/assets/732025/8094162/87322948-0fc5-11e5-8ae4-cadb6e5653c3.png>
> truecrypt_003 <https://cloud.githubusercontent.com/assets/732025/8094164/87345b82-0fc5-11e5-86d5-0104aa393303.png>
>
> |$ opensc-tool -a
> Using reader with a card: Yubico Yubikey NEO OTP+CCID 00 00
> |
>
> Further reading:
> http://forum.yubico.com/viewtopic.php?p=7166&sid=958cdb2de8b74e2320955704d3765a12#p7166
> http://www.gossamer-threads.com/lists/gnupg/users/70510
> http://sourceforge.net/p/veracrypt/discussion/features/thread/34fb4681/
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/479>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
Looking closer at the VeraCrypt code, it uses "KeyFiles" that contain keying material, not a specific type of key.

When used from disk:
    https://veracrypt.codeplex.com/wikipage?title=Keyfiles%20in%20VeraCrypt
says:
	"Any kind of file (for example, .txt, .exe, mp3**, .avi) can be used as a VeraCrypt keyfile (however, we recommend
          that you prefer compressed files, such as .mp3, .jpg, .zip, etc)."

When uses with PKCS#11 it appears the VeraCrypt code is looking for private CKO_DATA files not CKO_SECRET_KEY keys.

The PIV specifications define a number of private (i.e. require PIN) data files:
"Cardholder Fingerprints(6010)", "Printed Information(3010)" and "Cardholder Iris Images (1015)".

All PIV objects on the card are in BER-TLV and the OpenSC code uses the outer TLV length to determine the size of the object.
It is not clear if the card actually parses the object when written or read.

The Yubikey NEO is not a full implementation of the PIV standards:

   https://developers.yubico.com/yubico-piv-tool/YubiKey_NEO_PIV_introduction.html

The NEO has a 2005 limit on object size, does not say anything about the above three objects,
or if they do support them if they treat them as private. (There is also an issue with NFC access to the card where the
PIV standards do not allow) The only data object the NEO talks abount is the CHUID, that is not private.

Therefore it might be possible to a write one of the private objects as long as at least the outer TLV is given
and matches object size.

https://www.opensc-project.org/opensc/wiki/PivTool

    PIV_EXT_AUTH_KEY="010203040506070801020304050607080102030405060708"
#i.e. the default key from NEO web page, if you changed it, use the new versions

    piv-tool -A M:9B:03 -O 6030 -in infile.der

infile.der is a binary file, with the outsize tag of 0x5C.


[So you could do this, but I would not trust my data to this hacked together solution.]

	












On 6/10/2015 4:10 PM,  ǝʞɔoʃʞ uɐʇsıɹʇ wrote:
> Disk encryption <https://github.com/OpenSC/OpenSC/wiki/Using-smart-cards-with-applications#disk-encryption>
>
>     TrueCrypt can use PKCS#11 tokens as keyfile stores. NB! TrueCrypt does not use asymmetric keys generated on the card but stores symmetric keys as data files in the token! This requires write
>     access to the token and keyfiles are extracted in plaintext on every use.
>
> Quote from Yubico/ykneo-openpgp#17 (comment) <https://github.com/Yubico/ykneo-openpgp/issues/17#issuecomment-94499111>
>
>     Hello,
>     Now that I got a Yubikey Neo last generation, I would like to use the PIV applet with OpenSC for PKCS11 authentication into VeraCrypt (fork of TrueCrypt) to force the need of a Yubikey to decrypt
>     my Volume.
>     My Yubikey & PIV applet is recognized by the OpenSC library that I specify in VeraCrypt.
>     However, when I try to use the opensc-pkcs11 library with Yubico inside VeraCrypt, I have an error saying "function not supported". I understand the problem come from OpenSC as it doesn't support
>     objects like keyfiles.
>
> security token keyfiles_001 <https://cloud.githubusercontent.com/assets/732025/8094163/873385c2-0fc5-11e5-89e2-cd8578d044b5.png>
> new security token keyfile properties_002 <https://cloud.githubusercontent.com/assets/732025/8094162/87322948-0fc5-11e5-8ae4-cadb6e5653c3.png>
> truecrypt_003 <https://cloud.githubusercontent.com/assets/732025/8094164/87345b82-0fc5-11e5-86d5-0104aa393303.png>
>
> |$ opensc-tool -a
> Using reader with a card: Yubico Yubikey NEO OTP+CCID 00 00
> |
>
> Further reading:
> http://forum.yubico.com/viewtopic.php?p=7166&sid=958cdb2de8b74e2320955704d3765a12#p7166
> http://www.gossamer-threads.com/lists/gnupg/users/70510
> http://sourceforge.net/p/veracrypt/discussion/features/thread/34fb4681/
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/479>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
Correction, the outer tag is 53, not 5C.

Here is an example of a demo fingerprint object from a NIST test card read from the card using

pkcs11-tool --slot 1 --module=/path/to/opensc-pkcs11.so -r -y data --application-id 2.16.840.1.101.3.7.2.96.16 --login -o fingerprints

$ od -t x1 < fingerprints
0000000 53 82 05 8d bc 82 05 87 03 0d 00 00 02 48 02 e7
0000020 00 1b 02 01 14 0a 0a 01 04 1e 00 5a 14 0a 0a 01
0000040 04 1e 00 5a 14 1e 0a 01 04 1e 00 5a 00 00 08 80
0000060 fe 4e 49 53 54 20 43 72 65 61 74 6f 72 00 00 00
0000100 00 00 00 d6 50 18 58 28 9d 6d ca cc 93 25 a1 68
[ parts removed ]
0002540 18 88 31 0c d5 5f 22 6c 50 af 07 93 39 45 fa 89
0002560 8a a4 98 2c 4f a2 2d 69 68 81 5d 9d 17 3a df 24
0002600 0f 3f 02 1c 5a c2 43 b9 d9 85 c2 6a 23 a5 f3 fe
0002620 00
0002621



-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by tristan-k** [comment]
Thanks. I'm not sure if I can follow your instructions. I'll report back here,

**Comment by frankmorgner** [comment]
@tristan-k any updates on this?

**Comment by tristan-k** [comment]
@frankmorgner I'm sorry havent got time for it.

**Comment by frankmorgner** [comment]
tl;dr:

OpenSC doesn't work with Truecrypt or its derivates. We will leave this issue as "won't fix", because
- unresolved licensing/support/security situation of Truecrypt
- currently nobody is willing to fix the situation within OpenSC

Feel free to open a PR if you have a fix (or re-open this issue if you have new input)

**Comment by WebArtsAG** [comment]
Hi,
as the TrueCrypt successor VeryCrypt is maintained and also supports pkcs11 to store key files the same way TrueCrypt did, if would be helpful to make this run with openSC. I tried, but openSC does not seem to support importing or exporting data, even not from/to the above three objects.

And I also cannot figure out how to check this from command line. What is an outer tag, how do I generate a file with an "outer tag" of 53, and how long should a file be? I know how to generate a file that contains random data with dd and /dev/random.

Thorsten

**Comment by mouse07410** [comment]
On the other hand, the way TrueCrypt and VeraCrypt appear to use PKCS11 tokens seems very <put your favorite derogatory term here>.

The main advantage of such a token is its ability to perform cryptographic operations, such as generating (pseudo) one-time pad/code **and signing or decrypting using asymmetric crypto**. To employ such a device as a glorified storage is IMHO beyond contempt.

May I suggest enhancing VeraCrypt functionality to support using PKCS11 devices as **cryptographic** processors rather than high-end USB sticks?

**Comment by WebArtsAG** [comment]
Hi mouse07410, yes I totally agree with you.

Perhaps a possible way would be be to encrypt the key (which is used to decrypt the volume header) newly salted every time, and use the smartcard functions to decrypt it and instantly re-encrypt it again with new salt, before even mounting the volume. The current version should not only be stored with the volume file, because an attacker with write access to the file would be able to overwrite the header with an old version.

If done properly, it would be more secure than an USB stick, because intercepting the communication between the token and the computer would not be very helpful for an attacker. He would need to compromise a version of the software that is used by an authorized person with present token. An unprivileged user would not be able to steal an uncrypted version of the key, because he would not be given write access to the software.

**Comment by mouse07410** [comment]
Yes. I'd say - either just encrypt the key with a PKCS11 token like NEO (or another PIV card, or CAC), or use the encrypted/decrypted material as something to mix with the key obtained from password and/or other keyfiles. 

Either way (protect the volume key exclusively with a PKCS11 token, or use what the token protects as a part of the mix to arrive at the "actual" volume key) would be fine, IMHO. To maintain the flexibility and leave the choice with the user, I'd recommend allowing the user to select whether to keep the "whole key" PKCS11-protected (using its asymmetric keys, of course - **not** storing it there! :), or using PKCS11-protected value as a part of the key (e.g., to be XOR-ed with something password- and/or keyfile-derived).

**Comment by dengert** [comment]
On 11/18/2015 8:31 AM, WebArtsAG wrote:
> Hi,
> as the TrueCrypt successor VeryCrypt is maintained and also supports pkcs11 to store key files the same way TrueCrypt did, if would be helpful to make this run with openSC. I tried, but openSC does
> not seem to support importing or exporting data, even not from/to the above three objects.

The NEO has a PIV like applet and an OpenPGP applet.

The PIV specifications define a applet that is meant to be only readable via the user.

But Yubico and OpenSC do provide tools to write data if you know the symmetric key.
You could use one of the three objects that require the PIN

>
> And I also cannot figure out how to check this from command line. What is an outer tag, how do I generate a file with an "outer tag" of 53, and how long should a file be? I know how to generate a file
> that contains random data with dd and /dev/random.

the 53 is a tag in asn1.

Asm an example, the facialimage object from a NIST demo card is 9717 bytes long
and starts with:

0000000 53 82 25 f1 bc 82 25 eb 03 0d 00 00 22 ac 02 e7
0000020 00 1b 05 01 14 0a 0a 01 04 1e 00 5a 14 0a 0a 01
0000040 04 1e 00 5a 14 1e 0a 01 04 1e 00 5a 00 00 02 20
0000060 fe 4e 49 53 54 20 43 72 65 61 74 6f 72 00 00 00
0000100 00 00 00 d6 50 18 58 28 9d 6d ca cc 93 25 a1 68
0000120 59 a4 69 27 c9 d4 5c 86 50 18 43 e2 00 00 00 00
0000140 46 41 43 00 30 31 30 00 00 00 22 ac 00 01 00 00

0022720 e2 c3 f8 c9 87 09 75 11 25 a1 87 0f 2f f8 dc 88
0022740 23 40 18 da 6d c4 05 1c ae 97 54 64 48 c5 93 fc
0022760 87 2f 80 fe 00
0022765

53 is tag
  82 says a 2 byte length
  25 f1 is the length of data under this tag.

  bc is  next level tag. looks like it is Context-specific constructed UniversalString
   82 two byte length
     25 eb   is length

at the end there is
  FE tag
    00 is length

See NIST 800-73-3 Part1 table 12

0022765 octal is 9717 decimal
25f1 hex is 9713 decimal
1(tag) + 1(82) + 2(the length bytes) + 9713(length of data) = 9717



A script I use to read one of the three PIV objects that could be used for other purposes that require the PIN
has:

export MODULE=/opt/smartcard/lib/opensc-pkcs11.so
SLOT=1
P11="pkcs11-tool --slot $SLOT --module $MODULE"
PDA="$P11 -r -y data --application-id"

  $PDA 2.16.840.1.101.3.7.2.96.16 --login -o fingerprints
  $PDA 2.16.840.1.101.3.7.2.48.1  --login -o printedinfo
  $PDA 2.16.840.1.101.3.7.2.96.48 --login -o facialimage

OpenSC will only check that the date read starts with a tag of 53, and the lengths match. of the tag+tag_length +length matches the data read.

The NEO may have restrictions on the size of an object and may or may not
support these three objects and may or may not require the PIN. You will need to try.

>
> Thorsten
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/479#issuecomment-157728995>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
The question has been asked before. And I agree that was not the intent of a PIV. If the device is used with other middleware
it may not work, because the middleware may try and parse the ASN.1 and choke on it.

But one could actually put a facial image of the user on the card and use it as random data as long as the image is not save.

Remember the NEO does not follow all the NIST PIV standards and is not on the approved list. They are selling it for users
and picked the PIV  because Windows has built in drivers for PIV.

On 11/18/2015 3:22 PM, Mouse wrote:
> Doug, you're certainly correct.
>
> But I think that while it is technically possible - using a PKCS11 token (such a NEO) in such a way borders on criminal. :-)
>
> I strongly discourage anyone from attempting to support writing user data (such as TrueCrypt does) to a PIV card.
>
> —
> Reply to this email directly or view it on GitHub <https://github.com/OpenSC/OpenSC/issues/479#issuecomment-157867784>.
>

-- 

  Douglas E. Engert  <DEEngert@gmail.com>

**Comment by mouse07410** [comment]
> But one could actually put a facial image of the user on the card and use it as random data
> as long as the image is not save.

Oh, absolutely. That facial image can be used as one of the keyfiles in TrueCrypt/VeraCrypt, sure. Except that I think writing that make to the card should be once-in-three-years operation. I'm pretty certain it can be done now with `piv-tool`. 

> Remember the NEO does not follow all the NIST PIV standards and is not on the approved list.

Yeah, but it does most of what I need a PIV token to do. And hopefully it will improve its conformance. And we might improve its conformance ourselves, by filling in the data objects that Yubico failed to supply.

> They are selling it for users and picked the PIV  because Windows has built in drivers for PIV.

:-)

Now with our little CCC trick, other middleware treats NEO as a real PIV most of the time.

**Comment by WebArtsAG** [comment]
Hi everyone, thanks for the useful info and the discussion :-)

1.) Yubico´s own PIV tool "yubico-piv-tool" seems not to contain functionality for writing facial image, fingerprint or printed information to the card. I will try using the OpenSC piv-tool, thank you @dengert for the detailed explanation.

2.) @mouse07410, I just started my research two weeks ago, because a client of us insists on using a token together with a trustworthy filesystem encryption, and they also want us to destroy the keys needed to access copies of the data that might reside on laptops, backup media or elsewhere after the project. They don´t trust closed-source encryption like FileVault. So after some research, I came up with VeraCrypt, as I read it supports PCKS11 tokens, and I woke up after digging deeper into it.

I will now contact the VeraCrypt developers, maybe we could even contribute there. I think it should be easily possible to utilize asymmetric encryption instead of file storage. Even a simple implementation would be secure enough for protection against normal users' tendency to sacrifice security for comfort – better than a USB key does.

In my opinion, a mature, really secure implementation must somehow use new data every time, to ensure the decrypted data is unusable for an attacker.

**Comment by mouse07410** [comment]
@WebArtsAG 

1. Any update? Particularly regarding VeraCrypt developers being open (or not) to the idea of supporting cryptographic tokens (like Yubikey) to wrap volume master key using PIV applet keys (or OpenPGP keys)?

2. Using fingerprint or facial image stored on a PIV token as one of the keyfiles is a fine idea, IMHO. I still think that (1) above is of a higher value and priority, but adding support for this wouldn't hurt.

**Comment by asbaklm** [comment]
It is possible to write a truecrypt/veracrypt keyfile as a data object to a pkcs11 token such as the Aventra MyEID using veracrypt or to import the keyfile via OpenSC and to protect access to that keyfile through the smartcard PIN.

Provided the card was finalised in OpenSC after it was initialised (important step), the PIN will be required to extract / read that data object, meaning that the truecrypt keyfile, stored as a data object on the card is unreadable unless the card PIN is supplied to make it available. 

Once the PIN has been entered the data object can be read, so the session ought to be closed after the TC volume has been mounted to prevent unwanted potential access to the keyfile.

It isn't ideal because the data object is exposed (and not protected against extraction and copying like a private key is) until the session has been closed, but it does work with Truecrypt's Keyfiles / Token files mechanism.

Combining this with a password to mount TC / Veracrypt volumes you have somewhat better protection than through the use of passwords alone. I don't know whether it's possible to store and PIN protect data objects on yubikeys.

**Comment by mouse07410** [comment]
> It is possible to write a truecrypt/veracrypt keyfile as a data object to a pkcs11 token

This is not what I asked about, and IMHO very bad design approach. As I said several times already, many tokens do **not** allow writing data objects, and for a good reason (say, security). So relying on the ability to write a keyfile to a token is a sign of a bad design.

**Comment by dengert** [comment]
Have you asked Yubico? They have a number of applets for the NEO,
    but only 2 supported by OpenSC, PIV and openpgp.
    The PIV standards do not let a user write files to the PIV applet.
    Their yubico-piv-tool does.  I don't know if the openpgp applet
    would work for you. 
    
    On 7/27/2016 12:20 PM, asbaklm wrote:
    
    
      It is possible to write a truecrypt/veracrypt keyfile as a data
        object to a pkcs11 token such as the Aventra MyEID using
        veracrypt or to import the keyfile via OpenSC and to protect
        access to that keyfile through the smartcard PIN.
      Provided the card was finalised in OpenSC after it was
        initialised (important step), the PIN will be required to
        extract / read that data object, meaning that the truecrypt
        keyfile, stored as a data object on the card is unreadable
        unless the card PIN is supplied to make it available. 
      Once the PIN has been entered the data object can be read, so
        the session ought to be closed after the TC volume has been
        mounted to prevent unwanted potential access to the keyfile.
      It isn't ideal because the data object is exposed (and not
        protected against extraction and copying like a private key is)
        until the session has been closed, but it does work with
        Truecrypt's Keyfiles / Token files mechanism.
      Combining this with a password to mount TC / Veracrypt volumes
        you have somewhat better protection than through the use of
        passwords alone. I don't know whether it's possible to store and
        PIN protect data objects on yubikeys.
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
    
    
    -- 

 Douglas E. Engert  <DEEngert@gmail.com>

**Comment by mouse07410** [comment]
> Have you asked Yubico?

Asked about what?

> They have a number of applets for the NEO, but only 2 supported by OpenSC, PIV and openpgp.

I think the only two applets that can be accessed this way are OpenPGP and PIV. VeraCrypt and TrueCrypt are using OpenPGP.

> The PIV standards do not let a user write files to the PIV applet.

And for a good reason.

> Their yubico-piv-tool does.

I believe it does when you accompany the command with a token management key.

Could you remind me how I can test it? A command to write a simple object, and a command to delete it.

> I don't know if the openpgp applet would work for you.

Frankly, I *don't want* it to work. It would be a security violation by design. If they want to write keyfiles to USB tokens - let them use USB sticks.

**Comment by dengert** [comment]
On 7/27/2016 8:19 PM, Mouse wrote:
    
    
      
        Have you asked Yubico?
      
      Asked about what?
    
    
    Maybe Yubiko could develop another applet. 
    
    
    
      
        They have a number of applets for the NEO, but only 2
          supported by OpenSC, PIV and openpgp.
      
      I think the only two applets that can be accessed this way are
        OpenPGP and PIV. VeraCrypt and TrueCrypt are using OpenPGP.
      
        The PIV standards do not let a user write files to the PIV
          applet.
      
      And for a good reason.
      
        Their yubico-piv-tool does.
      
      I believe it does when you accompany the command with a token
        management key.
      Could you remind me how I can test it? A command to write a
        simple object, and a command to delete it.
    
    
    See my replies on this issue dated 6/12/15 above.  
    https://github.com/OpenSC/OpenSC/issues/479#issuecomment-110982730
    https://github.com/OpenSC/OpenSC/issues/479#issuecomment-111541479
    
      
        I don't know if the openpgp applet would work for you.
      
      Frankly, I don't want it to work. It would be a
        security violation by design. If they want to write keyfiles to
        USB tokens - let them use USB sticks.
    
    
    Then don't use it, I would not use it either. But @asbaklm asked if
    it could be done. 
    
    
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
    
    
    -- 

 Douglas E. Engert  <DEEngert@gmail.com>

**Comment by mouse07410** [comment]
> Maybe Yubiko could develop another applet.

Maybe - but it would be rather stupid of them to do so, and probably could jeopardize certifications for normal use (aka PIV and OpenPGP).

Of course the OP is free to ask them.

>> I believe [yubico-piv-tool] does [allow writing to the card] when you accompany the command with a token management key. Could you remind me how I can test it? A command to write a simple object, and a command to delete it.
>
> See my replies on this issue dated 6/12/15 above. 

Ah, I see, thanks. It appears to end up requiring the card management key (that symmetric key you were talking about). Which is how it's supposed to be, AFAICT. Whoever provisions the card (right to do so is enforced by the knowledge of the symmetric card management key), can put whatever object(s) he sees fit. Card's regular user can read the objects - some with PIN, and some without, and of course perform crypto operations with the card.

See also my reply https://github.com/OpenSC/OpenSC/issues/479#issuecomment-157867784  :-)

>> Frankly, I *don't want it* to work. It would be a security violation by design. If they want to write keyfiles to USB tokens - let them use USB sticks.
>
> Then don't use it, I would not use it either. But @asbaklm asked if it could be done.

Ah, I see. Yes, you're correct, of course. But to me it's like asking "can I run a car into a wall at 60 kph?" The answer it "yes you can - but..."

**Comment by dengert** [comment]
On 7/28/2016 6:46 AM, Mouse wrote:
    
    
      
        Maybe Yubiko could develop another applet.
      
      Maybe - but it would be rather stupid of them to do so, and
        probably could jeopardize certifications for normal use (aka PIV
        and OpenPGP).
    
    
    Certification??? By  who? Not by NIST
    
 http://csrc.nist.gov/groups/SNS/piv/npivp/validation_lists/PIVCardApplicationValidationList.htm
    
    Yubico took the specs and developed their own version of an applet.
    
    
    (OpenSC is not certificated middleware either.) 
    
    
    
      Of course the OP is free to ask them.
      
        
          I believe [yubico-piv-tool] does [allow writing to the
            card] when you accompany the command with a token management
            key. Could you remind me how I can test it? A command to
            write a simple object, and a command to delete it.
        
        See my replies on this issue dated 6/12/15 above. 
      
      Ah, I see, thanks. It appears to end up requiring the card
        management key (that symmetric key you were talking about).
        Which is how it's supposed to be, AFAICT. Whoever provisions the
        card (right to do so is enforced by the knowledge of the
        symmetric card management key), can put whatever object(s) he
        sees fit. Card's regular user can read the objects - some with
        PIN, and some without, and of course perform crypto operations
        with the card.
      See also my reply #479
          (comment) :-)
      
        
          Frankly, I don't want it to work. It would be a
            security violation by design. If they want to write keyfiles
            to USB tokens - let them use USB sticks.
        
        Then don't use it, I would not use it either. But @asbaklm asked if it could be done.
      
      Ah, I see. Yes, you're correct, of course. But to me it's like
        asking "can I run a car into a wall at 60 kph?" The answer it
        "yes you can - but..."
    
    
    Mythbusters!
    
    http://www.discovery.com/tv-shows/mythbusters/videos/100-mph-crash/
    
    
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
    
    
    -- 

 Douglas E. Engert  <DEEngert@gmail.com>

**Comment by dengert** [comment]
@asbaklm  asked:
It is possible to write a truecrypt/veracrypt keyfile as a data object to a pkcs11 token such as the Aventra MyEID using veracrypt or to import the keyfile via OpenSC and to protect access to that keyfile through the smartcard PIN.

I nave no idea what the MyEID card can do. @mouse07410 and I have been talking about the NEO, and both agree that its not a good idea.

**Comment by mouse07410** [comment]
> It is possible to write a truecrypt/veracrypt keyfile as a data object to a pkcs11 token such as the Aventura MyEID

The answer depends on (a) the specific card - whether its applet (JavaCard in the MyEID case) would allow that in general, and (b) how that card/applet is configured - whether it's been provisioned to allow such operations or not. Any sane configuration would prohibit the user from writing data objects to the token, for many reasons. 

[Adentra MyEID](https://services.aventra.fi/pdf/ActiveSecurity%20MyEID%20Cards%20white%20paper%20(2p)%20EN.pdf) is a Finnish PIV-compatible JavaCard that apparently allows creation of a file system:

> In addition, it is possible to create a custom file structure for generic applications. MyEID is also PIV compatible, and all the interfaces share the same data and file structure.

It is unclear what the above quote means - that every user can write data objects to it as files, or that the provisioning authority can put data objects on it prior to releasing the card to the user. 

> [We] have been talking about the NEO, and both agree that its not a good idea.

Yes we both agree. However what we said does not seem to be limited to NEO, and should equally well apply to any smart card.

**Comment by dengert** [comment]
Not much technical docs. 
https://services.aventra.fi/pdf/ActiveSecurity%20MyEID%20Cards%20white%20paper%20(2p)%20EN.pdf
    references  opensc-project.org  that is no longer
    correct. 
    
    If anyone has an existing MyEID card, that works with OpenSC, to
    test if it has a VIA applet, try this command:
    opensc-tool -s "00 A4 04 00 09 A0 00 00 03 08 00 00 10 00"
    If it returns 90 00 it has a PIV applet. 
    
    But the PIV specs define all the objects the applet can use. You
    cant add new objects and be considered PIV. 
    
    On 7/29/2016 1:56 PM, Mouse wrote:
    
    
      
        It is possible to write a truecrypt/veracrypt keyfile as a
          data object to a pkcs11 token such as the Aventura MyEID
      
      The answer depends on (a) the specific card - whether its
        applet (JavaCard in the MyEID case) would allow that in general,
        and (b) how that card/applet is configured - whether it's been
        provisioned to allow such operations or not. Any sane
        configuration would prohibit the user from writing data objects
        to the token, for many reasons. 
      Adentra
          MyEID is a Finnish PIV-compatible JavaCard that apparently
        allows creation of a file system:
      
        In addition, it is possible to create a custom file structure
          for generic applications. MyEID is also PIV compatible, and
          all the interfaces share the same data and file structure.
      
      It is unclear what the above quote means - that every user can
        write data objects to it as files, or that the provisioning
        authority can put data objects on it prior to releasing the card
        to the user. 
      
        [We] have been talking about the NEO, and both agree that its
          not a good idea.
      
      Yes we both agree. However what we said does not seem to be
        limited to NEO, and should equally well apply to any smart card.
      —
        You are receiving this because you were mentioned.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
    
    
    -- 

 Douglas E. Engert  <DEEngert@gmail.com>

**Comment by asbaklm** [comment]
"I nave no idea what the MyEID card can do."

Apparently not. But I did provide you a clue as to one of the things it can do.

" @mouse07410 and I have been talking about the NEO, and both agree that its not a good idea. "

That's very nice but what you and mouse get up to in your spare time is not necessarily of interest or relevance to the entire universe.

The only reason I commented was because I was interested at the time in whether or not the Yubico could provide some kind of improved solution on my already existing one, which I'm fully aware is not particularly great. Still, it beats having a password only or a keyfile on a usb stick. (Again, I am *aware* that it's by no means ideal.)

Some of the OTHER THAN YOURSELF readers may find it interesting, considering that for better or worse, that's how Truecrypt / Veracrypt stores keyfiles on certain types of pkcs11 cards. Ideally they should rewrite Truecrypt / Veracrypt to use the private key on a pkcs11 instead but for now that doesn't appear to be on the cards.
